server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -252,28 +258,79 @@ private void loadSnapshotInfos(SnapshotsInProgress snapshotsInProgress, String r
                 snapshotSet.add(new SnapshotInfo(entry));
             }
         }
-        // then, look in the repository
+        // then, look in the repository if there's any matching snapshots left
+        final List<SnapshotInfo> snapshotInfos;
+        if (snapshotIdsToIterate.isEmpty()) {
+            snapshotInfos = Collections.emptyList();
+        } else {
+            snapshotInfos = Collections.synchronizedList(new ArrayList<>());
+        }
+        final ActionListener<Collection<Void>> allDoneListener = listener.delegateFailure((l, v) -> {
+            final ArrayList<SnapshotInfo> snapshotList = new ArrayList<>(snapshotInfos);
+            snapshotList.addAll(snapshotSet);
+            CollectionUtil.timSort(snapshotList);
+            listener.onResponse(unmodifiableList(snapshotList));
+        });
+        if (snapshotIdsToIterate.isEmpty()) {
+            allDoneListener.onResponse(Collections.emptyList());
+            return;
+        }
+        // put snapshot info downloads into a task queue instead of pushing them all into the queue to not completely monopolize the
+        // snapshot meta pool for a single request
+        final int workers = Math.min(threadPool.info(ThreadPool.Names.SNAPSHOT_META).getMax(), snapshotIdsToIterate.size());
+        final Executor executor = threadPool.executor(ThreadPool.Names.SNAPSHOT_META);
```

Comment:
Used the same logic here that we use for file uploads and I believe also pre-warming where we do this kind of fake work-stealing-pool. This could be dried up in a follow-up since we have that same logic in a few places now.